src/starter.py

```diff
@@ -0,0 +1,211 @@
+
+import httpx
+from pydantic import BaseModel
+
+
+class Service(BaseModel):
+    cmd: List[str]
+    cwd: str
+    socket: Optional[str]
+    port: Optional[int]
+
+
+class RunConfig(BaseModel):
+    backend: Optional[Service]
+    frontend: Optional[Service]
+    nginx: Optional[Service]
+    svc_wait_time: int
+
+
+async def run(
+    cmd: List[str],
+    cwd: Optional[str] = None,
+    pipe_output=False,
+    extra_env: Optional[dict] = None,
+) -> Process:
+    """
+    Run a process
+    :param cmd: Command to run
+    :param cwd: Working directory
+    :param pipe_output: Whether to pipe the output or print it on screen instead
+    :param extra_env: Extra environment variables
+    :return: Process
+    """
+    env = os.environ.copy()
+    if extra_env:
+        env.update(extra_env)
+
+    proc = await asyncio.create_subprocess_exec(
+        cmd[0],
+        *cmd[1:],
+        stdout=asyncio.subprocess.PIPE if pipe_output else sys.stdout,
+        stderr=asyncio.subprocess.PIPE if pipe_output else sys.stderr,
+        cwd=cwd,
+        env=env,
+    )
+
+    return proc
+
+
+def debug_msg(msg: str) -> None:
+    """
+    Print message only when DEBUG environment variable is set
+    :param msg: Message to print
+    """
+    if os.environ.get("DEBUG"):
+        print(msg)
+
+
+async def run_and_wait(cmd: List[str], cwd: Optional[str] = None):
+    proc = await run(cmd, cwd)
+    code = await proc.wait()
+    if code != 0:
+        raise RuntimeError(f"{cmd} failed with code {code}")
+
+
+async def get_output(cmd: List[str], cwd: Optional[str] = None) -> bytes:
+    """
+    Run command and get its stdout if completed successfully
+    :param cmd: Command to run
+    :param cwd: Working directory
+    :return: Bytes from stdout
+    """
+    proc = await run(cmd, cwd, pipe_output=True)
+    out, err = await proc.communicate()
+    if proc.returncode != 0:
+        raise RuntimeError(f"{cmd} failed with code {proc.returncode}")
+    return out
+
+
+async def stop_process(proc: Process, timeout: float = 3.0) -> None:
+    """
+    Try to stop the process gracefully, otherwise killing it
+    :param proc: Process to stop
+    :param timeout: How long to wait for a graceful stop
+    """
+    if proc.returncode is not None:
+        return
+    proc.terminate()
+    await asyncio.wait_for(proc.wait(), timeout=timeout)
+    if proc.returncode is not None:
+        return
+    else:
+        proc.kill()
+
+
+async def setup_nginx() -> None:
+    """
+    Prepare Nginx.
+    Optionally setup HTTP Basic Auth, if environment variables are provided.
+    Process Nginx templates to build actual configuration files.
+    """
+    if user := os.environ.get("HTTP_BASIC_AUTH_USER"):
+        pwd = os.environ["HTTP_BASIC_AUTH_PASSWORD"]
+        await run_and_wait(["htpasswd", "-b", "-c", "/run/nginx/.htpasswd", user, pwd])
+        if not os.environ.get("HTTP_BASIC_AUTH_REALM"):
+            os.environ["HTTP_BASIC_AUTH_REALM"] = "Private area"
```

Comment:
How about instead of setting this to the env here passing it on as an exta env thing to `get_output` and let it pass on to `run` that does take in extra env stuff.